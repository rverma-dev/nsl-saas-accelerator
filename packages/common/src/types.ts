import TIER from './stacks.json';

export interface DeploymentRecord {
  tenantId?: string; // callhealth
  id: string; // dev1, dev2, prod1, prod2
  type?: string; // can be pool or silo
  tier?: string; // small
  account?: string;
  region?: string;
}

export interface Deployment extends DeploymentRecord {
  provisioned?: boolean;
}
export function isValidTier(tierName: string): boolean {
  return TIER.some(tier => tier.name === tierName);
}

export function getPipelineName(record: DeploymentRecord): string {
  return `${record.tenantId}-${record.id}-${record.type}-pipeline`;
}
