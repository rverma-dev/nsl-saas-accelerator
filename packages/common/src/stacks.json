[
  {
    "name": "small",
    "eks": {
      "version": "1.25"
    },
    "msk": {
      "version": "2.8"
    },
    "rds": {
      "version": "10.x",
      "size": "db.small"
    },
    "dynamodb": {
      "size": "14-2"
    }
  },
  {
    "name": "small-rc",
    "eks": {
      "version": "1.25"
    },
    "msk": {
      "version": "2.8"
    },
    "rds": {
      "version": "10.x",
      "size": "db.small"
    },
    "dynamodb": {
      "size": "14-2"
    }
  },
  {
    "name": "large",
    "eks": {
      "version": "1.25"
    },
    "msk": {
      "version": "2.8"
    },
    "rds": {
      "version": "10.x",
      "size": "db.small"
    },
    "dynamodb": {
      "size": "100-80"
    }
  },
  {
    "name": "serverless",
    "eks": {
      "version": "1.25"
    },
    "msk": {
      "version": "2.8"
    },
    "rds": {
      "version": "10.x",
      "maxSize": "2"
    },
    "dynamodb": {
      "maxSize": "14"
    }
  }
]