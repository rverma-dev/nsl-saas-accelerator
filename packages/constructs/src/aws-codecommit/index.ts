/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import { RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface CreateGitopsSecretProps {
  /**
   * Gitops IAM username
   */
  username: string;
  /**
   * Secret name
   */
  secretName: string;
}

/**
 * Class to configure CloudWatch Destination on logs receiving account
 */
export class CreateGitopsSecretResource extends Construct {
  readonly id: string;

  constructor(scope: Construct, id: string, props: CreateGitopsSecretProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::CodeCommitCredentials';
    const accessRoleResourceName = 'CreateGitopsSecretLambda';
    const assetsAccessRole = new cdk.aws_iam.Role(this, accessRoleResourceName, {
      roleName: `AWSLambdaBasicExecutionRole-${this.node.id}`,
      assumedBy: new cdk.aws_iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'AWS Accelerator assets access role in workload accounts to bootstrap gitops credentials',
    });
    assetsAccessRole.addManagedPolicy(
      cdk.aws_iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
    );
    assetsAccessRole.addToPolicy(
      new cdk.aws_iam.PolicyStatement({
        resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:*`],
        actions: ['logs:CreateLogGroup'],
      }),
    );
    assetsAccessRole.addToPolicy(
      new cdk.aws_iam.PolicyStatement({
        resources: [
          `arn:${cdk.Aws.PARTITION}:secretsmanager:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:secret:${props.secretName}`,
        ],
        actions: ['secretsmanager:*'],
      }),
    );
    assetsAccessRole.addToPolicy(
      new cdk.aws_iam.PolicyStatement({
        resources: ['*'],
        actions: ['kms:Decrypt', 'kms:DescribeKey'],
      }),
    );
    assetsAccessRole.addToPolicy(
      new cdk.aws_iam.PolicyStatement({
        resources: [`arn:${cdk.Aws.PARTITION}:iam::${cdk.Aws.ACCOUNT_ID}:user/${props.username}`],
        actions: [
          'iam:CreateServiceSpecificCredential',
          'iam:DeleteServiceSpecificCredential',
          'iam:ListServiceSpecificCredentials',
        ],
      }),
    );
    //
    // Function definition for the custom resource
    //
    const providerLambda = new cdk.aws_lambda.Function(this, 'Function', {
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, '../../lib/aws-codecommit/lambda-codecommit-secret')),
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(15),
      description: 'Create Gitops Secret',
      role: assetsAccessRole,
    });

    // Custom resource lambda log group
    new cdk.aws_logs.LogGroup(this, `${providerLambda.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${providerLambda.functionName}`,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const provider = new cdk.custom_resources.Provider(this, 'Custom::GitopsSecrets', {
      onEventHandler: providerLambda,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        username: props.username,
        secretName: props.secretName,
      },
    });

    this.id = resource.ref;
  }
}
