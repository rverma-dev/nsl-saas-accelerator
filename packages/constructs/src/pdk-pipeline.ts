import { Aspects, RemovalPolicy, SecretValue, Stage } from 'aws-cdk-lib';
import { Pipeline } from 'aws-cdk-lib/aws-codepipeline';
import { GitHubTrigger } from 'aws-cdk-lib/aws-codepipeline-actions';
import { Key } from 'aws-cdk-lib/aws-kms';
import { BlockPublicAccess, Bucket, BucketEncryption, ObjectOwnership } from 'aws-cdk-lib/aws-s3';
import * as pipelines from 'aws-cdk-lib/pipelines';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

const DEFAULT_BRANCH_NAME = 'main';
const REPOSITORY_SECRET = 'saas-provisoner';

/**
 * Properties to configure the PDKPipeline.
 *
 * Note: Due to limitations with JSII and generic support it should be noted that
 * the synth, synthShellStepPartialProps.input and
 * synthShellStepPartialProps.primaryOutputDirectory properties will be ignored
 * if passed in to this construct.
 *
 * synthShellStepPartialProps.commands is marked as a required field, however
 * if you pass in [] the default commands of this construct will be retained.
 */
export interface PDKPipelineProps extends pipelines.CodePipelineProps {
  /**
   * Name of the CodeCommit repository to create.
   */
  readonly repositoryName: string;

  /**
   * Output directory for cdk synthesized artifacts i.e: packages/infra/cdk.out.
   */
  readonly primarySynthDirectory: string;

  /**
   * PDKPipeline by default assumes a NX Monorepo structure for it's codebase and
   * uses sane defaults for the install and run commands. To override these defaults
   * and/or provide additional inputs, specify env settings, etc you can provide
   * a partial ShellStepProps.
   */
  readonly synthShellStepPartialProps?: pipelines.ShellStepProps;

  /**
   * Branch to trigger the pipeline execution.
   *
   * @default mainline
   */
  readonly defaultBranchName?: string;

  /**
   * Alias to use for existing KMS Key when crossAccount.
   *
   * @default mainline
   */
  readonly existingKMSKeyAlias?: string;

  /**
   * Possible values for a resource's Removal Policy
   * The removal policy controls what happens to the resource if it stops being managed by CloudFormation.
   */
  readonly codeCommitRemovalPolicy?: RemovalPolicy;

  /**
   * If we want to reuse existing artifact bucket
   * However for cross account access its better to create bucket per tenant
   */
  readonly existingAccessLogBucket?: string;

  /**
   * If we want to reuse existing artifact bucket
   * However for cross account access its better to create bucket per tenant
   */
  readonly existingArtifactBucket?: string;

  /**
   * Determine if pipeline is for toolchain or workload
   */
  readonly isToolChain?: string;
}

/**
 * An extension to CodePipeline which configures sane defaults for a NX Monorepo
 * codebase. In addition to this, it also creates a CodeCommit repository with
 * automated PR builds and approvals.
 */
export class SaasPipeline extends Construct {
  readonly codePipeline: pipelines.CodePipeline;

  public constructor(scope: Construct, id: string, props: PDKPipelineProps) {
    super(scope, id);

    this.node.setContext('@aws-cdk/aws-s3:serverAccessLogsUseBucketPolicy', true);

    const accessLogsBucket = props.existingAccessLogBucket ?
      Bucket.fromBucketName(this, 'ExAccessLogsBucket', props.existingAccessLogBucket) :
      new Bucket(this, 'AccessLogsBucket', {
        versioned: false,
        enforceSSL: true,
        autoDeleteObjects: true,
        removalPolicy: RemovalPolicy.DESTROY,
        encryption: BucketEncryption.S3_MANAGED,
        objectOwnership: ObjectOwnership.BUCKET_OWNER_ENFORCED,
        publicReadAccess: false,
        blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      });

    const artifactBucket = props.existingArtifactBucket ?
      Bucket.fromBucketName(this, 'ExArtifactsBucket', props.existingArtifactBucket) :
      new Bucket(this, 'ArtifactsBucket', {
        enforceSSL: true,
        autoDeleteObjects: true,
        removalPolicy: RemovalPolicy.DESTROY,
        encryption: props.crossAccountKeys ? BucketEncryption.KMS : BucketEncryption.S3_MANAGED,
        encryptionKey: props.crossAccountKeys && props.existingKMSKeyAlias ?
          Key.fromLookup(this, 'ArtifactKey', { aliasName: props.existingKMSKeyAlias }) :
          props.crossAccountKeys ?
            new Key(this, 'ArtifactKey', {
              enableKeyRotation: true,
              removalPolicy: RemovalPolicy.DESTROY,
            }) : undefined,
        objectOwnership: ObjectOwnership.BUCKET_OWNER_ENFORCED,
        publicReadAccess: false,
        blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
        serverAccessLogsPrefix: 'access-logs',
        serverAccessLogsBucket: accessLogsBucket,
      });

    const githubInput = pipelines.CodePipelineSource.gitHub(
      props.repositoryName,
      props.defaultBranchName || DEFAULT_BRANCH_NAME,
      {
        trigger: GitHubTrigger.NONE,
        authentication: SecretValue.secretsManager(REPOSITORY_SECRET, {
          jsonField: 'github_token',
        }),
      },
    );

    const codePipeline = new Pipeline(this, 'CodePipeline', {
      enableKeyRotation: props.crossAccountKeys,
      restartExecutionOnUpdate: true,
      crossAccountKeys: props.crossAccountKeys,
      artifactBucket,
    });

    // ignore input and primaryOutputDirectory
    const { input, primaryOutputDirectory, commands, ...synthShellStepPartialProps } =
    props.synthShellStepPartialProps || {};

    const synthShellStep = new pipelines.ShellStep('Synth', {
      input: githubInput,
      installCommands: ['npm install -g aws-cdk pnpm', 'pnpm install --frozen-lockfile'],
      commands: commands && commands.length > 0 ? commands : ['npx nx run-many --target=build --all'],
      primaryOutputDirectory: props.primarySynthDirectory,
      ...(synthShellStepPartialProps || {}),
    });

    synthShellStep.addOutputDirectory('.');

    const codePipelineProps: pipelines.CodePipelineProps = {
      codePipeline,
      ...props,
      crossAccountKeys: undefined,
      synth: synthShellStep,
    };

    this.codePipeline = new pipelines.CodePipeline(this, id, codePipelineProps);
  }

  /**
   * @inheritDoc
   */
  addStage(stage: Stage, options?: pipelines.AddStageOpts): pipelines.StageDeployment {
    // Add any root Aspects to the stage level as currently this doesn't happen automatically
    Aspects.of(stage.node.root).all.forEach(aspect => Aspects.of(stage).add(aspect));
    return this.codePipeline.addStage(stage, options);
  }

  addWave(wave: string, options?: pipelines.WaveOptions): pipelines.Wave {
    return this.codePipeline.addWave(wave, options);
  }

  buildPipeline() {
    this.codePipeline.buildPipeline();
    this.suppressCDKViolations();
  }

  suppressCDKViolations() {
    this.suppressRules(
      ['AwsSolutions-IAM5', 'AwsPrototyping-IAMNoWildcardPermissions'],
      'Wildcards are needed for dynamically created resources.',
    );

    this.suppressRules(
      ['AwsSolutions-CB4', 'AwsPrototyping-CodeBuildProjectKMSEncryptedArtifacts'],
      'Encryption of Codebuild is not required.',
    );

    this.suppressRules(
      ['AwsSolutions-S1', 'AwsPrototyping-S3BucketLoggingEnabled'],
      'Access Log buckets should not have s3 bucket logging',
    );
  }

  private suppressRules(rules: string[], reason: string) {
    NagSuppressions.addResourceSuppressions(
      this,
      rules.map(r => ({
        id: r,
        reason,
      })),
      true,
    );
  }
}
