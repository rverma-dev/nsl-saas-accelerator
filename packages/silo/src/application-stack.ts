import { EksCluster } from '@nsa/construct';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class ApplicationStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    new EksCluster(this, {
      platformTeamRole: scope.node.tryGetContext('PLATFORM_TEAM_ROLE')!,
      vpcID: scope.node.tryGetContext('VPC_ID'),
    });
  }
}