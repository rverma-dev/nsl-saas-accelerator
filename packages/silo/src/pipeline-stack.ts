import { DeploymentRecord } from '@nsa/common';
import { SaasPipeline } from '@nsa/construct';
import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ApplicationStage } from './application-stage';

interface WorkloadPipelineProps extends DeploymentRecord {
  readonly toolchainKms?: string;
  readonly toolchainLogBucket?: string;
  readonly toolchainAssetBucket?: string;
}

export class PipelineStack extends Stack {
  readonly pipeline: SaasPipeline;

  constructor(scope: Construct, id: string, props: WorkloadPipelineProps) {
    super(scope, id, { env: { account: props.account, region: props.region } });
    this.pipeline = new SaasPipeline(this, `${props.tenantId}-${props.id}-pool`, {
      primarySynthDirectory: 'packages/pool/cdk.out',
      repositoryName: this.node.tryGetContext('repositoryName') || 'rverma-nsl/nsl-saas-accelerator',
      publishAssetsInParallel: false,
      crossAccountKeys: true,
      synth: {},
      dockerEnabledForSynth: true,
      existingKMSKeyAlias: props.toolchainKms,
      existingArtifactBucket: props.toolchainAssetBucket,
      existingAccessLogBucket: props.toolchainLogBucket,
    });
    const devStage = new ApplicationStage(this, 'Dev', { env: { account: props.account, region: props.region } });
    this.pipeline.addWave('application').addStage(devStage);
    this.pipeline.buildPipeline(); // Needed for CDK Nag
  }
}
