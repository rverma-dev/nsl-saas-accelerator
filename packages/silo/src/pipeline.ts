import { EksCluster } from '@nsa/construct';
import { aws_dynamodb, Fn, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode } from 'aws-cdk-lib/aws-dynamodb';
import { PDKNag } from 'aws-prototyping-sdk/pdk-nag';
import { Construct } from 'constructs';

export class Silo extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const eks = new EksCluster(this, {
      platformTeamRole: scope.node.tryGetContext('PLATFORM_TEAM_ROLE')!,
      vpcID: scope.node.tryGetContext('VPC_ID'),
      gitopsRepoBranch: scope.node.tryGetContext('GITOPS_REPO_BRANCH'),
      gitopsRepoUrl: Fn.importValue('CodeCommitRepoUrlExport'),
      gitopsRepoSecret: Fn.importValue('CodeCommitSecretNameExport'),
    });
    const dynamo = new aws_dynamodb.Table(this, 'dynamo', {
      partitionKey: { name: 'id', type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
    });
    eks.node.addDependency(dynamo);
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = PDKNag.app();

new Silo(app, 'silo-dev', { env: devEnv });

app.synth();
