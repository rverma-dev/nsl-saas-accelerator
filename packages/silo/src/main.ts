import { PDKNag } from '@aws-prototyping-sdk/pdk-nag';
import { PipelineStack } from './pipeline-stack';

const app = PDKNag.app();
const tenantProps = {
  tenantId: 'CH',
  id: 'dev-001',
  type: 'pool',
  tier: 'small',
  account: process.env.CDK_DEFAULT_ACCOUNT!,
  region: process.env.CDK_DEFAULT_REGION!,
};

new PipelineStack(app, 'PipelineStack', tenantProps);
app.synth();
