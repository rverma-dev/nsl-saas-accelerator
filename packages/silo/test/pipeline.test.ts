import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Silo } from '../src/pipeline';

test('Snapshot', () => {
  const app = new App();
  app.node.setContext('PLATFORM_TEAM_ROLE', 'Admin');
  app.node.setContext('VPC_ID', 'vpc-12345678');
  const stack = new Silo(app, 'test', {
    env: {
      account: '123456789',
      region: 'us-east-1',
    },
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});