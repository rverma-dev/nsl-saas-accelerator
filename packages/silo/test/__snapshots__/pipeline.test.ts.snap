// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "dynamo3C1BB246": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
